### DISCUSSION TITLE
Keyboard shortcuts

### DISCUSSION THREAD

**Discussion by RickCogley** [retypeapp/retype#214] [Ideas]
Hitting `/` puts the cursor in search, and I thought it would also be cool to have something mapped to those automatically generated previous and next links.

**Comment by geoffreymcgill** [comment]
Agreed. Excellent idea.

I will investigate and try to figure out what keys would work best. My initial thought would be `[` and `]` keys.

**Reply by larmarange** [reply]
Such characters could be  difficult to type on non English keyboards.

**Reply by larmarange** [reply]
Letters n and p are common in many web applications

**Reply by RickCogley** [reply]
On a Japanese keyb, if you are switched to English mode, the 「 and 」 keys map to [ and ]. FYI. 

Another idea is simply to map to arrow keys. There are either physical keys or shortcuts for those usually.

**Reply by geoffreymcgill** [reply]
Gmail uses `j` and `k` to move through older and newer conversations.

Action | Shortcut
-- | --
Older conversation | j
Newer conversation | k

Reddit uses `p` and `n` keys to navigate between posts:

Action | Shortcut
-- | --
Previous post in lightbox | p
Next post in lightbox | n

I would likely stay away from the left|right arrow keys as they likely have native functionality attached, such as scroll viewport left or right.

Seems like `p` for `previous` and `n` for `next` are good fit. Thoughts?

**Reply by DannyBen** [reply]
> Seems like `p` for `previous` and `n` for `next` are good fit. Thoughts?

I fell like these keys need to be a) together, and b) in the correct direction.

1. I have seen `j` and `k` in other places as well. Also the indentation on the `j` key helps to use these two keys without looking (similarly to how your work with arrows).
2. I feel `p` and `n` are wrong. I doubt these get any significant real user use. It might be ok for places like reddit, where Next is a more used function than Previous (I presume).
3. `b` and `n` might also be good (`b` for back). These pair are also closer to the bottom of the keyboard, which makes them feel closer to how the arrows feel.
4. Ideally, I would hope to have the arrows do that. The viewport should not have a horizontal scrollbar. If it does, something in the design is wrong. The only places that might have horizontal scrollbars are code areas. The way I handle it on my sites that have arrows, is to disable its navigation action if a scrollable component is active.
5. Another killer feature that arrows have, is the fact that the other two arrows are right there at your fingertips - so you scroll up and down to read, and next and back to switch pages.

Another piece of information: [mdbook uses arrows](https://rust-lang.github.io/mdBook/), and it is pretty natural and straight forward.

**Reply by geoffreymcgill** [reply]
Using `j` and `k` or arrow keys would avoid confusion with the blog pages because the buttons are `Newer` (left) and `Older` (right). On regular content pages, it would be `Previous` (left) and `Next` (right). 

The arrows are a good fit. 

It would be relatively trivial for us to make these keyboard shortcuts configurable, although just landing on good solution that most everyone can agree on would be preferred.

**Reply by DannyBen** [reply]
> although just landing on good solution that most everyone can agree on would be preferred.

Well, this is a hard thing to do in software design. Personally, I tend to choose the intuitive approach for things like this, and in this case, the choice is even clearer since there are keys designed *specifically* for this purpose in the keyboard. 

I think that if you allow configuration (or disabling at least) of these keys, then there is no other key pair with a stronger argument than arrow keys. Even without configuration, I feel arrow keys have more advantages than disadvantages.

**Reply by geoffreymcgill** [reply]
Left and right arrows it is. We will try and get this feature into the `v1.12` release.

**Reply by RickCogley** [reply]
cool, that will be a nice UX win

**Reply by geoffreymcgill** [reply]
We are reviewing this feature request and are now undecided on whether this is a good idea or not, especially if using the `left-arrow` and `right-arrow` keys as navigation. Those keys are assigned to horizontal scrolling in basically all browsers and operating systems. 

Overriding the native functionality may have unintended consequences that we just cannot foresee at the moment. I would be most concerned with impacting accessibility browsers and/or users that need to zoom their content for visual impairment.

Thoughts?

**Reply by DannyBen** [reply]
> Thoughts?

I for one am happy with any of the below:

1. No implementation at all. Reason: I do not expect web pages to have any "built in" navigation.
2. Have just hotkeys provided by the native [`accesskey`](https://www.w3schools.com/tags/att_global_accesskey.asp) functionality, so on Windows+Chrome it would be the <kbd>ALT</kbd> key + whatever keys are configured (<kbd>ALT</kbd>+<kbd>n</kbd> for next maybe, and <kbd>ALT</kbd>+<kbd>b</kbd>for back?)
3. Have arrow keys, but in a way that a) they can be disabled in config and b) they are not hijacking any normal functionality (Vertical scroll, if exists, search field etc).
4. How do you feel about <kbd>ALT</kbd>+<kbd>→</kbd> and <kbd>ALT</kbd>+<kbd>←</kbd>? I have seen these used in some web interfaces, and logically, it makes sense: ALTernative use for the arrow keys.

**Comment by spencercap** [comment]
[ meta + K ] is also a common key command for focusing the search input

**Reply by spencercap** [reply]
how do you specify a key modifier for the hot keys config in yaml?

**Reply by geoffreymcgill** [reply]
> how do you specify a key modifier for the hot keys config in yaml?

Retype does not include configurable keyboard shortcut options.

**Reply by spencercap** [reply]
dang. 

i got so close manually inserting some javascript for the keydown event listener but it seems you have setup a more complex isFocussed state using vue reactivity. any chance you can expose these reactive state models to a global variable off of window.searchState or similar?
